streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryWindowStore.java

```diff
@@ -0,0 +1,398 @@
+                final WrappedK<K> wrappedKey = kvMapEntry.getKey();
+                returnSet.add(getWindowedKeyValue(wrappedKey.getKey(), segmentMapEntry.getKey(),
+                    kvMapEntry.getValue()));
+            }
+        }
+        return new InMemoryWindowedKeyValueIterator<>(returnSet.listIterator());
+    }
+
+    @Override
+    public boolean persistent() {
+        return false;
+    }
+
+    @Override
+    public boolean isOpen() {
+        return this.open;
+    }
+
+    @Override
+    public void flush() {
+        // do-nothing since it is in-memory
+    }
+
+    @Override
+    public void close() {
+        this.segmentMap.clear();
+        this.open = false;
+    }
+
+    private List<KeyValue<Long, V>> fetchUnique(final K key, final long timeFrom, final long timeTo) {
+        final List<KeyValue<Long, V>> returnSet = new LinkedList<>();
+
+        // add one b/c records expire exactly retentionPeriod ms after created
+        final long minTime = max(timeFrom, this.context.streamTime() - this.retentionPeriod + 1);
+
+        for (final Map.Entry<Long, NavigableMap<WrappedK<K>, V>> segmentMapEntry : this.segmentMap.subMap(minTime, true, timeTo, true).entrySet()) {
+            final V value = segmentMapEntry.getValue().get(new WrappedK<>(key, seqnum));
+            if (value != null) {
+                returnSet.add(new KeyValue<>(segmentMapEntry.getKey(), value));
+            }
+        }
+        return returnSet;
+    }
+
+    private List<KeyValue<Long, V>> fetchWithDuplicates(final K key, final long timeFrom, final long timeTo) {
+        final List<KeyValue<Long, V>> returnSet = new LinkedList<>();
+
+        // add one b/c records expire exactly retentionPeriod ms after created
+        final long minTime = max(timeFrom, this.context.streamTime() - this.retentionPeriod + 1);
+        final WrappedK<K> keyFrom = new WrappedK<>(key, 0);
+        final WrappedK<K> keyTo = new WrappedK<>(key, Integer.MAX_VALUE);
+
+        for (final Map.Entry<Long, NavigableMap<WrappedK<K>, V>> segmentMapEntry : this.segmentMap.subMap(minTime, true, timeTo, true).entrySet()) {
+            for (final Map.Entry<WrappedK<K>, V> kvMapEntry : segmentMapEntry.getValue().subMap(keyFrom, true, keyTo, true).entrySet()) {
+                returnSet.add(new KeyValue<>(segmentMapEntry.getKey(), kvMapEntry.getValue()));
+            }
+        }
+        return returnSet;
+    }
+
+    private void removeExpiredSegments() {
+        final long minLiveTime = this.context.streamTime() - this.retentionPeriod;
+        final NavigableMap<Long, NavigableMap<WrappedK<K>, V>> expiredSegments = this.segmentMap.headMap(minLiveTime, true);
+        for (Iterator<Entry<Long, NavigableMap<WrappedK<K>, V>>> it = expiredSegments.entrySet().iterator(); it.hasNext(); ) {
+            it.next();
+            it.remove();
+        }
+    }
+
+    private KeyValue<Windowed<K>, V> getWindowedKeyValue(final K key, final long startTimestamp, final V value) {
```

Comment:
Not a comment for this PR: when we remove the generics, this function can be replaced with `WindowStoreIteratorWrapper` since the inner iterator would always be bytes.